packages/pkl.experimental.structuredRead/structuredRead.pkl

```diff
@@ -0,0 +1,261 @@
+          deepToTyped.apply(targetClass, mappedInputs)
+        else
+          mappedInputs
+
+/// Sets the scheme used with `read()`. By default is `prop:` so values
+/// are read from external properties passed using the `-p` flag.
+///
+/// If set to `env:` this would result in values being read from
+/// environment variables.
+hidden inputScheme: String(matches(Regex(#"^[\d\w-]+:$"#)))
+
+/// Path seperator used to when computing the `read()` URI of a nested
+/// value. By default `.` which means that nested values are passed in
+/// using a dot seperated path as the key value.
+hidden pathSeperator: String
+
+/// Template used to convert a path of elements into a complete URI to
+/// be passed to `read?()`
+hidden pathTemplate: (List<String>) -> String = (path: List<String>) -> "\(inputScheme)\(path.join(pathSeperator))"
+
+/// Function used to read in values. Can be overwritten to change how
+/// values are read. Useful for testing
+hidden readFunc = (s) -> read?(s)
+
+function applyToClass(target: Class, path: List<String>): Mapping<String, Any?> =
+  let (reflected = reflect.Class(target))
+  new Mapping {
+    for (name, property in reflected.properties) {
+      when (isCoerceable(property)) {
+        when (isNullableProp(property)) {
+          [name] = let (potentialValue = readValueToTarget(path, name, property)) if (!(potentialValue is ReadFailure)) potentialValue else property.defaultValue
+        } else {
+          [name] = let (potentialValue = readValueToTarget(path, name, property)) if (!(potentialValue is ReadFailure)) potentialValue else property.defaultValue ?? potentialValue
+        }
+      }
+    }
+  }
+
+local function isCoerceable(prop: reflect.Property) =
+  let (_ = prop.type.getClass().simpleName)
+  (prop.type is reflect.DeclaredType) ||
+    (prop.type is reflect.NullableType) ||
+    (prop.type is reflect.TypeAlias) ||
+    (prop.type is reflect.UnknownType)
+local function isNullableProp(prop: reflect.Property) =
+  prop.type is reflect.NullableType
+
+local function throwFailures(target: Class, fails: List<Failure>) =
+  throw(
+    let (msg =
+      if (fails.length > 1)
+        new Listing {
+          "Error processing properties in \(target)):\n"
+          for (fail in fails) {
+            renderFail(fail, fails)
+          }
+        }
+      else
+        new Listing {
+          "Error processing property in \(target):\n"
+          renderFail(fails[0], fails)
+        }
+    ) msg.join("\n")
+  )
+
+local function renderFail(fail: Failure, fails: List<Failure>): String =
+  let (path = fail.path.join(pathSeperator))
+    let (longestPath = longestPath(fails))
+      let (padding = " ".repeat(longestPath - path.length))
+        "\t'\(path)'\(padding): \(fail.message)"
+
+local function longestPath(fails: List<Failure>) =
+  fails.fold(0, (acc, f) ->
+    let (path = f.path.join(pathSeperator))
+      if (path.length > acc)
+        path.length
+      else acc
+  )
```

Comment:
I believe this is "just"
```
fails.map((f) -> f.path.join(pathSeperator).length).max
```
and can be inlined. It also helps not to do this for _every_ `fail` in `fails`, so I also suggest inlining this in `renderFail`, inlining `renderFail` into `assetNoFailures`;
```
local function assertNoFailures(target: Class, output: Mapping<String, Any?>|Dynamic): Boolean =
  let (fails = findFailures(output))
    fails.isEmpty ||
      let (msg = new Listing {
        "Error processing \(if (fails.length > 1) "properties" else "property") in \(target)):\n"
        local maxPathLength = fails.map((f) -> f.path.join(pathSeperator).length).max
        for (fail in fails) {
          let (path = fail.path.join(pathSeperator))
            "\t'\(path)'\(" ".repeat(maxPathLength - path.length)): \(fail.message)"
        }
      }.join("\n"))
        throw(msg)
```

(If I hadn't prematurely submitted the first half of this review, I would have inlined. Soz)

Reply:
Yes it does appear to be "just" 

```
fails.map((f) -> f.path.join(pathSeperator).length).max
```

looks like my functional foo failed me here